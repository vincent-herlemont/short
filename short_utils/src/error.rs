//! short specific Error type
//! Allow to embedded multiple error type.
use super::result::Result;
use promptly::ReadlineError;
/// Inspiration from : https://github.com/brson/basic-http-server/blob/1ab052719a88e41822b2955d7d72bf161457d47c/src/main.rs#L468
use serde::export::fmt::Debug;
use serde_yaml;

use std::error::Error as StdError;
use std::fmt::Result as FmtResult;
use std::fmt::{Display, Formatter};
use std::io;
use std::path::StripPrefixError;

use std::string::FromUtf8Error;

pub type ExitStatusCode = Option<i32>;

// TODO: Add PartialEq to Error type.
#[derive(Debug)]
pub enum Error {
    Other(String),
    Wrap(String, Box<Error>),
    Io(io::Error),
    SerdeYaml(serde_yaml::Error),
    Which(which::Error),
    StripPrefixError(StripPrefixError),
    ExitStatus(ExitStatusCode),
    RustyLine(ReadlineError),
    StringUtf8(FromUtf8Error),
}

impl Error {
    /// Create new error
    /// # Example
    /// Usually use [`format!`] as string parameter for pass context information.
    ///
    /// ```
    /// use short_utils::error::Error;
    /// let url = "http://.....";
    /// Error::new(format!("fail to get url {}",url));
    /// ```
    pub fn new<S: AsRef<str>>(msg: S) -> Error {
        Error::Other(String::from(msg.as_ref()))
    }

    /// Create new box error
    pub fn new_box(msg: String) -> Box<Error> {
        Box::new(Error::new(msg))
    }

    pub fn wrap<S: AsRef<str>>(msg: S, err: Error) -> Error {
        Error::Wrap(String::from(msg.as_ref()), Box::new(err))
    }

    pub fn is<P>(&self, predicate: P) -> bool
    where
        P: Fn(&Error) -> bool,
    {
        match self {
            Error::Wrap(_, err) => err.is(predicate),
            err => predicate(err),
        }
    }
}

impl Display for Error {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        match self {
            Error::Other(msg) => write!(f, "{}", msg),
            Error::Wrap(msg, err) => write!(f, "{} : {}", msg, err),
            Error::Io(err) => write!(f, "{}", err),
            _ => write!(f, "{:?}", self),
        }
    }
}

impl StdError for Error {
    fn source(&self) -> Option<&(dyn StdError + 'static)> {
        use Error::*;

        match self {
            Io(e) => Some(e),
            SerdeYaml(e) => Some(e),
            Which(_) => None,
            Wrap(_, error) => Some(error),
            Other(_) => None,
            StripPrefixError(e) => Some(e),
            ExitStatus(_) => None,
            RustyLine(e) => Some(e),
            StringUtf8(e) => Some(e),
        }
    }
}

impl From<&'static str> for Error {
    fn from(message: &'static str) -> Error {
        Error::Other(String::from(message))
    }
}

impl From<String> for Error {
    fn from(message: String) -> Error {
        Error::Other(message)
    }
}

impl From<io::Error> for Error {
    fn from(error: io::Error) -> Error {
        Error::Io(error)
    }
}

impl From<serde_yaml::Error> for Error {
    fn from(error: serde_yaml::Error) -> Error {
        Error::SerdeYaml(error)
    }
}

impl From<which::Error> for Error {
    fn from(error: which::Error) -> Error {
        Error::Which(error)
    }
}

impl From<StripPrefixError> for Error {
    fn from(error: StripPrefixError) -> Error {
        Error::StripPrefixError(error)
    }
}

impl From<ReadlineError> for Error {
    fn from(error: ReadlineError) -> Error {
        Error::RustyLine(error)
    }
}

impl From<ExitStatusCode> for Error {
    fn from(exit_status: ExitStatusCode) -> Error {
        Error::ExitStatus(exit_status)
    }
}

impl From<FromUtf8Error> for Error {
    fn from(error: FromUtf8Error) -> Error {
        Error::StringUtf8(error)
    }
}

impl Error {
    pub fn exit_code_eq(&self, code: i32) -> Result<bool> {
        if let Error::ExitStatus(exit_status) = self {
            let cmd_code = exit_status.ok_or(Error::new("Process terminated by signal"))?;
            return Ok(code == cmd_code);
        }
        Err(Error::new("no exit status")).unwrap()
    }
}

#[cfg(test)]
mod tests {
    use super::Error;
    use crate::result::Result;
    use std::io;
    use std::io::ErrorKind;

    #[test]
    fn error_test() {
        fn create_io_error() -> Result<()> {
            Err(Error::from(io::Error::new(io::ErrorKind::Other, "ho no !")))
        }
        fn create_other_error() -> Result<()> {
            Err(Error::new(format!("ho no !")))
        }
        fn create_other_error_from() -> Result<()> {
            Err(Error::from("ho no !"))
        }

        let result = create_io_error();
        match result.err().unwrap() {
            Error::Io(_) => assert!(true),
            _ => assert!(false),
        };

        let result = create_other_error();
        match result.err().unwrap() {
            Error::Other(_) => assert!(true),
            _ => assert!(false),
        };

        let result = create_other_error_from();
        match result.err().unwrap() {
            Error::Other(_) => assert!(true),
            _ => assert!(false),
        };
    }

    #[test]
    fn nested_error() {
        let err = Error::from(io::Error::new(ErrorKind::NotFound, "not found"));
        let err = Error::wrap("wrap lvl 1", err);
        let err = Error::wrap("wrap lvl 2", err);
        let b = err.is(|err| matches!(err, Error::Io(_)));
        assert!(b);
        let b = err.is(|err| matches!(err, Error::Other(_)));
        assert!(!b);
    }
}
