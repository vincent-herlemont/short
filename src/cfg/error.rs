use std::path::PathBuf;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum CfgError {
    #[error("setup not found `{0}`")]
    SetupNotFound(String),
    #[error("project not found `{0:?}`")]
    ProjectNotFound(PathBuf),
    #[error("project `{0:?}` already added")]
    ProjectAlreadyAdded(PathBuf),
    #[error("private env dir not found for `{0}`")]
    PrivateEnvDirNotFound(String),
    #[error("private env dir must be an absolute path `{0:?}` for `{1}`")]
    PrivateEnvDirRelativePath(PathBuf, String),
    #[error("public env dir not found for `{0}`")]
    PublicEnvDirNotFound(String),
    #[error("public env dir already unset for `{0}`")]
    PublicEnvAlreadyUnset(String),
    #[error("private env dir already unset for `{0}`")]
    PrivateEnvAlreadyUnset(String),
    #[error("unknown cfg error")]
    Unknown,
}
