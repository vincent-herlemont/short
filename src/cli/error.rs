use std::path::PathBuf;
use thiserror::Error;

type SetupName = String;

#[derive(Error, Debug)]
pub enum CliError {
    #[error("`{0:?}` not found for `{1}`")]
    EnvDirNotFound(PathBuf, SetupName, #[source] std::io::Error),
    #[error("open editor fail")]
    OpenEditorFail,
    #[error("bad input `{0}` try again")]
    ConfirmBadInputTryAgain(String),
    #[error("you have not allowed to delete var `{0}`:`{1}` in {2}")]
    DeleteVarNowAllowed(String, String, String),
    #[error("env must be sync, please change it manually or run \"short env sync\"")]
    EnvFileMustBeSync,
    #[error("env file `0:?` already exist")]
    EnvFileAlreadyExist(PathBuf),
}
