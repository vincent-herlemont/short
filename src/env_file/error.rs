use std::io;
use std::path::PathBuf;

use thiserror::Error;

#[derive(Error, Debug)]
pub enum EnvReaderError {
    #[error("io env reader error")]
    Io {
        #[from]
        source: io::Error,
    },
    #[error("space on var name `{0}`")]
    SpaceOnVarName(String),
    #[error("unknown env error")]
    Unknown,
}

#[derive(Error, Debug)]
pub enum EnvError {
    #[error("io env error")]
    Io {
        #[from]
        source: io::Error,
    },
    #[error("fs_extra env error")]
    FsExtra {
        #[from]
        source: fs_extra::error::Error,
    },
    #[error("fail to parse `{file:?}`")]
    FailToParse {
        #[source]
        source: EnvReaderError,
        file: PathBuf,
    },
    #[error("env var `{0}` not found in `{1:?}`")]
    EnvVarNotFound(String, PathBuf),
    #[error("env file `{0:?}` has no file name")]
    EnvFileHasNoFileName(PathBuf),
    #[error("env file `{0:?}` has an empty file name")]
    EnvFileNameIsEmpty(PathBuf),
    #[error("env file `{0:?}` has incorrect file name : it must begin with `.` char")]
    EnvFileNameIncorrect(PathBuf),
}
