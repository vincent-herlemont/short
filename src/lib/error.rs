//! D4d specific type of error

use serde::export::fmt::Debug;
use std::error::Error;
use std::fmt::{Display, Formatter, Result};

#[derive(Debug)]
pub struct D4dError {
    pub detail: String,
}

impl D4dError {
    /// Create new error
    /// # Example
    /// Usually use [`format!`] as string parameter for pass context information.
    ///
    /// ```
    /// let url = "http://.....";
    /// D4dError::new(format!("fail to get url",url);
    /// ```
    pub fn new(msg: String) -> D4dError {
        D4dError { detail: msg }
    }

    /// Create new box error
    pub fn new_box(msg: String) -> Box<D4dError> {
        Box::new(D4dError::new(msg))
    }
}

impl Display for D4dError {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        write!(f, "{}", self.detail)
    }
}

impl Error for D4dError {}
