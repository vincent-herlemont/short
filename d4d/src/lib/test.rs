//! Helper for test related of d4d domain.
use crate::lib::resource;
use crate::lib::resource::Resource;
use std::fs;
use std::path::PathBuf;
use tempdir::TempDir;

pub struct Config {
    pub tmp_dir: PathBuf,
}

impl Drop for Config {
    fn drop(&mut self) {
        fs::remove_dir_all(self.tmp_dir.clone()).expect("can not clean tmp directory");
    }
}

/// Return [`InspectorConfig`], create temporary directory and copy resource on it.
///
/// The temporary directory is owned by [`InspectorConfig.path`].
///
/// # Recommendation
///
/// Need to call [`after`] at the end of test.
pub fn before(test_name: &str) -> Config {
    let test_name = format!("{}.{}", "d4d", test_name);

    // Create temporary directory.
    let path = TempDir::new(test_name.as_str())
        .expect("fail to create temporary directory")
        .into_path();

    // Copy resources to it.
    resource::to_dir(&path).expect("fail to copy resources");

    Config { tmp_dir: path }
}

/// Assert that patter value or|and expression is present on an vector.
///
/// # Notice
/// The macro don't take the ownership of a vector.
///
/// You have to add [`#[allow(unreachable_patterns)]`] to avoid warning
///
/// # Example
/// ```
/// let v = vec![1,2,3,4];
///assert_find!(v, 2);
///assert_find!(v, 8); // assertion failed: can not found {8}  in {v}
///assert_find!(v, el, el < &&5)
///assert_find!(v, el, el > &&5); // assertion failed: can not found {el} with expresion {el > &&5} in {v}
///
/// ```
#[macro_export]
macro_rules! assert_find {
    ($v:ident,$arms:pat) => {
        assert_find!($v, $arms, true, true)
    };
    ($v:ident,$arms:pat,$e:expr) => {
        assert_find!($v, $arms, $e, true)
    };
    ($v:ident,$arms:pat,$e:expr,$b:expr) => {
        if (&$v)
            .iter()
            .find(|el| match el {
                $arms => $e,
                _ => false,
            })
            .is_some()
            != $b
        {
            panic!(format!(
                "assertion failed: {} {{{}}} {} in {{{}}}",
                if ($b) {
                    String::from("can not found")
                } else {
                    String::from("found")
                },
                String::from(stringify!($arms)),
                if (stringify!($e) == "true" || stringify!($e) == "false") {
                    String::from("")
                } else {
                    String::from(format!(
                        "with expresion {{{}}}",
                        String::from(stringify!($e))
                    ))
                },
                String::from(stringify!($v))
            ));
        };
    };
}

/// Assert that patter value or|and expression is not present on an vector.
///
/// # Notice
/// The macro don't take the ownership of a vector.
///
/// You have to add [`#[allow(unreachable_patterns)]`] to avoid warning
///
/// # Example
/// ```
/// let v = vec![1,2,3,4];
///assert_not_find!(v, 8);
///assert_not_find!(v, 2); // assertion failed: found {2}  in {v}
///assert_not_find!(v, el, el > &&5);
///assert_not_find!(v, el, el < &&5); // assertion failed: found {el} with expresion {el < &&5} in {v}
/// ```
#[macro_export]
macro_rules! assert_not_find {
    ($v:ident,$arms:pat) => {
        assert_find!($v, $arms, true, false)
    };
    ($v:ident,$arms:pat,$e:expr) => {
        assert_find!($v, $arms, $e, false)
    };
}

#[cfg(test)]
mod tests {

    #[allow(unreachable_patterns)]
    #[test]
    fn assert_find_macro_test() {
        let v = vec![1, 2, 3, 4];
        assert_find!(v, 2);
        // assert_find!(v, 8); // assertion failed: can not found {8}  in {v}
        assert_find!(v, el, el < &&5);
        // assert_find!(v, el, el > &&5); // assertion failed: can not found {el} with expresion {el > &&5} in {v}

        assert_not_find!(v, 8);
        // assert_not_find!(v, 2); // assertion failed: found {2}  in {v}
        assert_not_find!(v, el, el > &&5);
        //assert_not_find!(v, el, el < &&5); // assertion failed: found {el} with expresion {el < &&5} in {v}
    }
}

pub const TEST_RESOURCE_DIRECTORY: &'static str = "./init_tpl";

/// Get all [`Resource`]
pub fn get_resource() -> Vec<Resource> {
    vec![
        Resource::new(
            "./init_tpl/other_conf.yaml",
            include_str!("./init_tpl/0_other_conf.yaml"),
        ),
        Resource::new(
            "./init_tpl/1_certificate.yaml",
            include_str!("./init_tpl/1_certificate.yaml"),
        ),
        Resource::new(
            "./init_tpl/1_certificate_altered.yaml",
            include_str!("./init_tpl/2_certificate_altered.yaml"),
        ),
        Resource::new(
            "./init_tpl/3_test/0_test.js",
            include_str!("./init_tpl/3_test/0_test.js"),
        ),
        Resource::new(
            "./init_tpl/4_tpl_certificate/certificate.yaml",
            include_str!("./init_tpl/4_tpl_certificate/certificate.yaml"),
        ),
    ]
}
