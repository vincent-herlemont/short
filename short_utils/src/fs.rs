//! File manipulation operations related of short domain.
use super::io;
use crate::error::Error;
use crate::result::{unwrap_partition, Result};
use std::cmp::Ordering;
use std::fs::File;
use std::io::BufReader;
use std::path::PathBuf;

/// SuperStructure including [`PathBuf`] of files and his content.
#[derive(Debug, Eq, PartialEq)]
pub struct ContentFile {
    pub path: PathBuf,
    pub contents: String,
}

impl Ord for ContentFile {
    fn cmp(&self, other: &Self) -> Ordering {
        self.path.cmp(&other.path)
    }
}

impl PartialOrd for ContentFile {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.path.cmp(&other.path))
    }
}

impl ContentFile {
    /// Return [`ContentFile`] when the predicate [`next_line`] is satisfied.
    pub fn read_contain<F>(path: &PathBuf, next_line: F) -> Result<ContentFile>
    where
        F: Fn(&str) -> bool,
    {
        let contents = match File::open(path).map(|file| BufReader::new(file)) {
            Ok(buffer) => io::read_to_string_contain(buffer, next_line),
            Err(error) => Err(Error::from(error)),
        };

        match contents {
            Ok(contents) => Ok(ContentFile {
                path: path.to_owned(),
                contents,
            }),
            Err(error) => Err(error),
        }
    }

    /// Return each [`ContentFile`] and [`Error`] when the predicate [`next_line`] is satisfied or not.
    pub fn read_contain_multi<F>(paths: &[PathBuf], next_line: F) -> (Vec<ContentFile>, Vec<Error>)
    where
        F: Fn(&str) -> bool,
    {
        let results: (Vec<_>, Vec<_>) = paths
            .into_iter()
            .map(|path| ContentFile::read_contain(path, &next_line))
            .partition(Result::is_ok);
        unwrap_partition(results)
    }
}

#[cfg(test)]
mod tests {
    use crate::asset::default_assets;
    use crate::assets::get_all;
    use crate::error::Error;
    use crate::fs::ContentFile;
    use crate::path::retrieve;
    use crate::test::before;
    use std::io;

    #[allow(unreachable_patterns)]
    #[test]
    fn read_contain_multi_test() {
        let config = before("search_test", default_assets(get_all()));
        let paths = retrieve(&config.tmp_dir).unwrap();
        let (mut content_files, errors) =
            ContentFile::read_contain_multi(&paths, |line| line.contains("test"));

        content_files.sort();
        // ContentFiles
        assert!((&content_files).len() >= 2);
        assert_find!(
            content_files,
            ContentFile { contents, path },
            contents == &String::from("console.log(\'test.js\');")
                && path.to_string_lossy().contains("test.js")
        );

        assert!((&errors).len() > 3);
        assert_find!(errors, Error::Io(e), e.kind() == io::ErrorKind::Other);
        assert_find!(errors, Error::Io(e), e.kind() == io::ErrorKind::NotFound);
    }
}
