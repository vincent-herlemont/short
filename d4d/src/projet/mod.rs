use std::fs::create_dir;
use std::path::Path;
use utils::error::Error;
use utils::result::Result;

#[allow(dead_code)]
const PROJECT_CONFIG_DIRECTORY: &'static str = ".d4d";

#[allow(dead_code)]
fn create_project_config_directory<P: AsRef<Path>>(home: P) -> Result<()> {
    let project_config_directory = home.as_ref().join(PROJECT_CONFIG_DIRECTORY);
    match create_dir(project_config_directory) {
        Ok(_) => Ok(()),
        Err(e) => Err(Error::from(e)),
    }
}

#[cfg(test)]
mod tests {
    use crate::projet::create_project_config_directory;
    use insta::assert_debug_snapshot;
    use std::collections::HashMap;
    use utils::test::before;
    use walkdir::WalkDir;

    #[test]
    fn test_create_project_config_directory() {
        let config = before("create_project_config_directory", HashMap::new());
        let r = create_project_config_directory(&config.tmp_dir);
        assert_debug_snapshot!(r);
        let r: Vec<_> = WalkDir::new(&config.tmp_dir).into_iter().collect();
        assert_debug_snapshot!(r.iter().count());
        let r = create_project_config_directory(&config.tmp_dir);
        assert_debug_snapshot!(r);
    }
}
