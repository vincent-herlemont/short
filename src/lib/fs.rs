//! File manipulation operations related of d4d domain.
use super::io;
use std::fs::File;
use std::io as sio;
use std::io::BufReader;
use std::path::PathBuf;

/// SuperStructure including [`File`] and his content.
#[derive(Debug)]
struct ContentFile {
    path: PathBuf,
    contents: String,
}

fn read_to_string_finds<F>(paths: &[PathBuf], f: F) -> (Vec<ContentFile>, Vec<sio::Error>)
where
    F: Fn(&str) -> bool,
{
    let (content_files, errors): (Vec<_>, Vec<_>) = paths
        .into_iter()
        .map(|path| {
            let contents = match File::open(path).map(|file| BufReader::new(file)) {
                Ok(buffer) => io::read_to_string_finds(buffer, &f),
                Err(error) => Err(error),
            };

            match contents {
                Ok(contents) => Ok(ContentFile {
                    path: path.to_owned(),
                    contents,
                }),
                Err(error) => Err(error),
            }
        })
        .partition(Result::is_ok);

    (
        content_files.into_iter().map(Result::unwrap).collect(),
        errors.into_iter().map(Result::unwrap_err).collect(),
    )
}

#[cfg(test)]
mod tests {
    use crate::lib::fs::read_to_string_finds;
    use crate::lib::path::retrieve;
    use crate::lib::test::{after, before};

    #[test]
    fn read_to_string_finds_test() {
        let tmpath = before("search_test");
        let paths = retrieve(&tmpath).unwrap();
        let (content_files, errors) = read_to_string_finds(&paths, |line| line.contains("test"));

        // ContentFiles
        assert_eq!((&content_files).len(), 1);
        assert_eq!((&content_files)[0].contents, "console.log(\'test.js\');");
        assert!((&content_files)[0]
            .path
            .to_str()
            .unwrap()
            .contains("test.js"));

        // Errors
        assert_eq!((&errors).len(), 2);
        after(tmpath);
    }
}
