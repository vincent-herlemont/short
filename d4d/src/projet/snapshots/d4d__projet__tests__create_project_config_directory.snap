---
source: d4d/src/projet/mod.rs
expression: r
---
Ok(
    (),
)
